bsp/src/org/jetbrains/bsp/BspServerWidgetProvider.scala

```diff
@@ -89,11 +98,54 @@ class BspServerWidgetProvider extends StatusBarWidgetProvider {
       }
     }
 
+    class TerminateServer extends AnAction(BspBundle.message("bsp.widget.stop.all.bsp.restart"), BspBundle.message("bsp.widget.stop.all.bsp.restart"), AllIcons.Actions.Exit) with DumbAware {
+
+      override def update(e: AnActionEvent): Unit = {
+        val workspace = new File(project.getBasePath)
+        val exitCommandDefined  = BspCommunicationService.getInstance.exitCommands(workspace.toURI).toOption.exists(_.nonEmpty)
```

Comment:
won't this do I/O on every update?

Reply:
No, because now the commands are fetched in `BspCommunication` constructor. I'm not fully sure it's valid but I could find any better way to implement `update` method. Please also note, I'm removing `BspConnection` from `conns` map on close.